packages/messaging/src/helpers/registerDefaultSw.ts

```diff
@@ -39,9 +39,40 @@ export async function registerDefaultSw(
     messaging.swRegistration.update().catch(() => {
       /* it is non blocking and we don't care if it failed */
     });
+    await waitForRegistrationActive(messaging.swRegistration);
   } catch (e) {
     throw ERROR_FACTORY.create(ErrorCode.FAILED_DEFAULT_REGISTRATION, {
       browserErrorMessage: (e as Error)?.message
     });
   }
 }
+
+/**
+ * Waits for registration to become active. MDN documentation claims that
+ * a service worker registration should be ready to use after awaiting
+ * navigator.serviceWorker.register() but that doesn't seem to be the case in
+ * practice, causing the SDK to throw errors when calling
+ * swRegistration.pushManager.subscribe() too soon after register(). The only
+ * solution seems to be waiting for the service worker registration `state`
+ * to become "active".
+ */
+async function waitForRegistrationActive(
+  registration: ServiceWorkerRegistration
+): Promise<void> {
+  return new Promise<void>((resolve, reject) => {
+    if (registration.active) {
+      resolve();
+    }
+    const incomingSw = registration.installing || registration.waiting;
+    if (incomingSw) {
+      incomingSw.onstatechange = ev => {
+        if ((ev.target as ServiceWorker)?.state === 'activated') {
+          incomingSw.onstatechange = null;
+          resolve();
+        }
+      };
+    } else {
+      reject(new Error('No incoming service worker found.'));
+    }
```

Comment:
I changed it to an if/else but not sure if this affects the timing at all.

Reply:
Yeah this is a tricky timing issue. I think now the SW could become active before we add the event listener, causing us to miss the `activated` event. For example:
1. registration is waiting
2. registration is not yet active, so we don't resolve immediately on line 79.
3. registration becomes active, and the `'activated'` event fires
4. attach registration state event listener
5. event listener never fires because we missed the `'activated'` event and we reject the promise after 10s

This is just me speculating, I'm not sure if this order of events is possible.